core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala

```diff
@@ -126,59 +149,91 @@ import com.datastax.driver.core.utils.UUIDs
   refreshInterval:          Option[FiniteDuration],
   bucketSize:               BucketSize,
   usingOffset:              Boolean,
-  initialTagPidSequenceNrs: Map[Tag, (TagPidSequenceNr, UUID)]
-) extends GraphStage[SourceShape[UUIDRow]] {
+  initialTagPidSequenceNrs: Map[Tag, (TagPidSequenceNr, UUID)]) extends GraphStage[SourceShape[UUIDRow]] {
 
   private val out: Outlet[UUIDRow] = Outlet("event.out")
+  private val verboseDebug = settings.eventsByTagDebug
 
   override def shape = SourceShape(out)
 
+  val stageUuid = UUID.randomUUID().toString
+
+  private def isLiveQuery(): Boolean = refreshInterval.isDefined
+
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
     new TimerGraphStageLogic(shape) with StageLogging with OutHandler {
       override protected def logSource = classOf[EventsByTagStage]
 
-      var currentOffset: UUID = fromOffset
-      var currTimeBucket: TimeBucket = TimeBucket(fromOffset, bucketSize)
-      var state: State = QueryIdle
-      var tagPidSequenceNrs: Map[Tag, (TagPidSequenceNr, UUID)] = initialTagPidSequenceNrs
-      var missingLookup: Option[LookingForMissing] = None
+      var stageState: StageState = _
+      val toOffsetMillis = toOffset.map(UUIDs.unixTimestamp).getOrElse(Long.MaxValue)
 
       lazy val system = materializer match {
         case a: ActorMaterializer => a.system
         case _ =>
           throw new IllegalStateException("EventsByTagStage requires ActorMaterializer")
       }
 
+      private def calculateToOffset(): UUID = {
+        val to: Long = UUIDs.unixTimestamp(UUIDs.timeBased()) - settings.eventsByTagEventualConsistency.toMillis
+        val tOff = if (to < toOffsetMillis) {
+          // The eventual consistency delay is before the end of the query
+          val u = UUIDs.endOf(to)
+          if (log.isDebugEnabled) {
+            log.debug("{}: New toOffset (EC): {}", stageUuid, formatOffset(u))
+          }
+          u
+        } else {
+          // The eventual consistency delay is after the end of the query
+          if (log.isDebugEnabled) {
+            log.debug("{}: New toOffset (End): {}", stageUuid, formatOffset(toOffset.get))
+          }
+          toOffset.get
+        }
+        tOff
+      }
+
+      private def updateToOffset(): Unit = {
+        updateStageState(_.copy(toOffset = calculateToOffset()))
+      }
+
+      private def updateQueryState(state: QueryState): Unit = {
+        updateStageState(_.copy(state = state))
+      }
+
       implicit def ec: ExecutionContextExecutor = materializer.executionContext
+
       setHandler(out, this)
 
       val newResultSetCb = getAsyncCallback[Try[ResultSet]] {
         case Success(rs) =>
-          val _ = state match {
+          val _ = stageState.state match {
```

Comment:
um, wat's up with the `val _`, wouldn't just doing the imperative stuffs in the ok match be more clean?

Reply:
What is up indeed!